Document virtualenv suppport for MLflow projects (#5727)
mlflow/mlflow: docs/source/projects.rst

  file in the root of the project directory or by including a ``conda_env`` entry in your
  ``MLproject`` file. For details, see the :ref:`project-directories` and :ref:`mlproject-specify-environment` sections.

>   The ``mlflow run`` command supports running a Conda environment project as a Python environment project.

Reviewer: Do we want to say "local Python environment" or leave it as "Python environment"?
I've thought of it as a hierarchy of 
Python Environment is a generic term implying that Python is available and installed/activated somewhere. 
That 'somewhere' could be:
Local, Virtualenv, Conda, or Docker, et al.

Author: > Python Environment is a generic term implying that Python is available and installed/activated somewhere.

Agree, what do you think about `Virtualenv project`?

Author: @dbczumar wdyt?

Author: One concern is matching to the `python_env` field.

- conda -> conda_env
- docker -> docker_env
- virtualenv -> python_env

Reviewer 2: `Virtualenv project` makes sense. `virtualenv -> python_env` is also just fine.

>   To do this, run ``mlflow run`` with ``--env-manager local``:

Author: ```suggestion
  To do this, run ``mlflow run`` with ``--env-manager virtualenv``:
```

  .. code-block:: bash

>       mlflow run /path/to/conda/project --env-manager virtualenv

Reviewer: should this be `--env-manager local` ?

Author: I think line 92 is incorrect :)

Reviewer: ah yes :)

  .. warning::

      When a Conda environment project is executed as a Python environment project,
      conda dependencies will be ignored and only pip dependencies will be installed.

.. _project-python-environments:

> Python environment

Author: ```suggestion
Virtualenv environment
```

Does `Virtualenv environment` sound strange because it repeats `env` twice?

Reviewer: It sounds fine to me since it's the name of a tool :)

  Python environments support Python packages available on PyPI. When an MLflow Project

>   specifies a Python environment, MLflow creates it using ``virtualenv`` and ``pyenv``, and

Reviewer: Do we want to say what these elements are doing? 
"MLflow will download the specified version of Python by using ``pyenv`` and create a ``virtualenv`` that contains the exact package dependencies and versions that were used during training, activating it as the execution environment prior to running the project code."
(or something like that)

Author: Sounds good! I'll add it.

Author: How about

> MLflow will download the specified version of Python by using pyenv and create an isolated environment that contains the project dependencies using virtualenv, activating it as the execution environment prior to running the project code.

?

Reviewer: sounds good!

  file in the root of the project directory or by including a ``conda_env`` entry in your
  ``MLproject`` file. For details, see the :ref:`project-directories` and :ref:`mlproject-specify-environment` sections.

>   The ``mlflow run`` command supports running a Conda environment project as a Virtualenv environment project.

Reviewer 2: Nit: `conda` is also lower-case; https://docs.conda.io/en/latest/

Reviewer: yep!

Author: @dbczumar @BenWilson2 Lower-cased

  To do this, run ``mlflow run`` with ``--env-manager virtualenv``:

  .. code-block:: bash

      mlflow run /path/to/conda/project --env-manager virtualenv

  .. warning::

      When a conda environment project is executed as a Virtualenv environment project,
      conda dependencies will be ignored and only pip dependencies will be installed.

.. _project-virtualenv-environments:

Virtualenv environment
  Virtualenv environments support Python packages available on PyPI. When an MLflow Project
  specifies a Virtualenv environment, MLflow will download the specified version of Python by using
  ``pyenv`` and create an isolated environment that contains the project dependencies using ``virtualenv``,
  activating it as the execution environment prior to running the project code.

>   You can specify a Virtualenv environment for your MLflow project by including a ``python_env`` entry in your

Reviewer 2: ```suggestion
  You can specify a Virtualenv environment for your MLflow Project by including a ``python_env`` entry in your
```

Author: Fixed!

  ``<MLFLOW_PROJECT_DIRECTORY>/files/config/conda_environment.yaml``, where
  ``<MLFLOW_PROJECT_DIRECTORY>`` is the path to the MLflow project's root directory.

Python environment
  Include a top-level ``python_env`` entry in the ``MLproject`` file.
  The value of this entry must be a *relative* path to a `python_env` YAML file
  within the MLflow project's directory. In the following example:

  .. code-block:: yaml

    python_env: files/config/python_env.yaml

  ``python_env`` refers to an environment file located at
  ``<MLFLOW_PROJECT_DIRECTORY>/files/config/python_env.yaml``, where
  ``<MLFLOW_PROJECT_DIRECTORY>`` is the path to the MLflow project's root directory.

>   The following is an example of an ``python_env.yaml`` file:

Author: ```suggestion
  The following is an example of a ``python_env.yaml`` file:
```

  ``<MLFLOW_PROJECT_DIRECTORY>/files/config/conda_environment.yaml``, where
  ``<MLFLOW_PROJECT_DIRECTORY>`` is the path to the MLflow project's root directory.

Virtualenv environment
  Include a top-level ``python_env`` entry in the ``MLproject`` file.
  The value of this entry must be a *relative* path to a `python_env` YAML file

>   within the MLflow project's directory. In the following example:

Reviewer 2: ```suggestion
  within the MLflow project's directory. The following is an example ``MLProject`` file with a ``python_env`` definition:
```

Author: Fixed!

    By default, MLflow Projects are run in the environment specified by the project directory
    or the ``MLproject`` file (see :ref:`Specifying Project Environments <project-environments>`).
    You can ignore a project's specified environment and run the project in the current

>     system environment by supplying the ``--env-manager=local`` flag.

Reviewer: Do you want to add a note or a warning here about ignoring version specifications can result in unstable or unpredictable prediction or inference results? Choosing to manually maintain the prediction environment and keeping it synchronized with the training environment is a bad idea without some form of environment automation and external handling of the CD process...

Author: Sounds good! I'll add it.
